routers/web/repo/view.go

```diff
@@ -1168,3 +1178,37 @@ func Forks(ctx *context.Context) {
 
 	ctx.HTML(http.StatusOK, tplForks)
 }
+
+func loadPinData(ctx *context.Context) error {
+	// First, cleanup any pins that are no longer valid
+	err := user_service.CleanupPins(ctx, ctx.Doer)
```

Comment:
Why cleanup pins when user home page be visited? Looks wired.

Reply:
We are cleaning the pins to avoid an edge case.
If we have a `user1`, with 6 pins, being one of those pins a repo belonging to `user2`. If `user 2` changes the repo visibility to private, `user1` can no longer have it as a pin. If `user1` now visited a new repo page, without visiting his own homepage, the `pin` button would be disabled because he already has the maximum number of pins. However, if we clean the pins, the pin to the repo of `user2` will be deleted, and the button will be enabled.

Reply:
When changing the visibility and user has no permission to access this pin, means user2 forcibly unpinned it, so this should be handled when `user2` changes the repo visibility, not the moment user visiting the repo home page.
Also for other cases, like remove the user as the member of an organization and so on.

Reply:
This is still a concern. Cleaning should not be done when a page is visiting.